lib/scholar/neighbors/graph.ex

```diff
@@ -0,0 +1,182 @@
+defmodule Scholar.Neighbors.Graph do
+  import Nx.Defn
+  require Nx
+  alias Scholar.Neighbors.RandomProjectionForest, as: Forest
+
+  knn_schema = [
+    k: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of neighbors in the graph."
+    ]
+  ]
+
+  # TODO: Rename this
+  approximate_knn_schema =
+    knn_schema ++
+      [
+        min_leaf_size: [
+          type: :pos_integer,
+          doc: "The minimum number of points in ..."
+        ],
+        num_trees: [
+          type: :pos_integer,
+          doc: "The number of trees in random projection forest."
+        ],
+        num_iters: [
+          type: :non_neg_integer,
+          default: 1,
+          doc: "The number of times to perform neighborhood expansion."
+        ],
+        key: [
+          type: {:custom, Scholar.Options, :key, []},
+          doc: """
+          Used for random number generation in parameter initialization.
+          If the key is not provided, it is set to `Nx.Random.key(System.system_time())`.
+          """
+        ]
+      ]
+
+  @knn_schema NimbleOptions.new!(knn_schema)
+  @approximate_knn_schema NimbleOptions.new!(approximate_knn_schema)
+
+  deftransform exact_brute_force(x, opts) do
+    opts = NimbleOptions.validate!(opts, @knn_schema)
+    exact_brute_force_n(x, opts)
+  end
+
+  defn exact_brute_force_n(x, opts) do
+    k = opts[:k]
+    # TODO: Add other metrics other than Euclidean
+    x
+    |> Nx.new_axis(1)
+    |> Nx.subtract(Nx.new_axis(x, 0))
+    |> Nx.pow(2)
+    |> Nx.sum(axes: [2])
+    |> Nx.put_diagonal(Nx.broadcast(:infinity, {Nx.axis_size(x, 0)}))
+    |> Nx.negate()
+    |> Nx.top_k(k: k)
+    |> elem(1)
+  end
+
+  @doc """
+  Constructs the approximate k-nearest neighbor graph using random projection forest
+  and neighborhood expansion.
+  """
+  deftransform approximate_random_projection_forest_and_expansion(x, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            """
+            expected input tensor to have shape {num_samples, num_features}, \
+            got tensor with shape: #{inspect(Nx.shape(x))}\
+            """
+    end
+
+    opts = NimbleOptions.validate!(opts, @approximate_knn_schema)
+    k = opts[:k]
+    min_leaf_size = opts[:min_leaf_size]
+
+    min_leaf_size =
+      if min_leaf_size do
+        if min_leaf_size > k, do: min_leaf_size, else: k + 1
+      else
+        max(10, 2 * k + 1)
+      end
+
+    size = Nx.axis_size(x, 0)
+    num_trees = if opts[:num_trees], do: opts[:num_trees], else: 5 + round(:math.pow(size, 0.25))
+    num_iters = opts[:num_iters]
+    key = Keyword.get_lazy(opts, :key, fn -> Nx.Random.key(System.system_time()) end)
+    forest = Forest.fit(x, num_trees: num_trees, min_leaf_size: min_leaf_size, key: key)
+    IO.inspect(forest.leaf_size)
+    IO.inspect(num_trees)
+    leaves = Forest.predict(forest, x) |> Nx.reshape({size, num_trees * forest.leaf_size})
```

Comment:
Perhaps if you convert this to:

```suggestion
    leaves = Forest.predict(forest, x) |> Nx.reshape({size, :auto})
```

Then you can say: give me any data structure that implements `predict`? Then it is up to you if you pass a random forest or kd-tree or whatever?

Reply:
The code would be something like:

```
tensor
|> Forest.fit()
|> NNExpand.fit()
```

Reply:
> The code would be something like:
> 
> ```
> tensor
> |> Forest.fit()
> |> NNExpansion.fit()
> ```

This could work, but then `Forest.fit/2` should return k-nearest neighbors or the graph and `NNExpansion.fit/2` should take a graph as input.

Reply:
Hrm, perhaps it doesn't make sense to think about how to abstract it unless we have another data structure we can use as the foundation for NNExpand. Maybe we keep this PR open until we have NNDescent or LargeVis or similar?

Reply:
> Hrm, perhaps it doesn't make sense to think about how to abstract it unless we have another data structure we can use as the foundation for NNExpand. Maybe we keep this PR open until we have NNDescent or LargeVis or similar?

Don't think having NNDescent or LargeVis is going to change it. LargeVis is basically random projection forest + NN-expansion - that is kinda what's done here.
What we could have instead is `Forest.fit_predict` (`fit_predict` is implemented in many scikit-learn modules). Now I am not sure what would I do with `NNExpansion`. It doesn't really go with fit/predict API.

Reply:
>  LargeVis is basically random projection forest + NN-expansion - that is kinda what's done here.

So to me that's perfect. We implement LargeVis and we embed this code inside. If someone wants to extract it out in the future for other use cases, we can do it once needed?

Reply:
> > LargeVis is basically random projection forest + NN-expansion - that is kinda what's done here.
> 
> So to me that's perfect. We implement LargeVis and we embed this code inside. If someone wants to extract it out in the future for other use cases, we can do it once needed?

Yeah, I think so. I don't really like the word LargeVis, especially since the algorithm seems natural to me (I had the same idea even before reading the paper), but then again the authors did it and not that I have a better name.

Reply:
At least is how people will recognize/search for it, so at least it helps on that aspect :)

Reply:
Agreed.